pkg/ottl/ottlfuncs/func_decode.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"golang.org/x/text/encoding/ianaindex"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type DecodeArguments[K any] struct {
+	Target   ottl.Getter[K]
+	Encoding string
+}
+
+func NewDecodeFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Decode", &DecodeArguments[K]{}, createDecodeFunction[K])
+}
+
+func createDecodeFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*DecodeArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("DecodeFactory args must be of type *DecodeArguments[K]")
+	}
+
+	return Decode(args.Target, args.Encoding)
+}
+
+func Decode[K any](target ottl.Getter[K], encoding string) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		val, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		var stringValue string
+
+		switch v := val.(type) {
+		case []byte:
```

Comment:
Could we also check for `pcommon.ByteSlice` and `pcommon.Value` that contains a string or byte slice?